src/Polysemy/Final.hs

```diff
@@ -36,107 +36,77 @@ module Polysemy.Final
--- to the final monad.
---
--- Consider using 'withStrategic' instead,
--- as it provides a more user-friendly interface to the same power.
---
--- You are discouraged from using 'withWeaving' directly in application code,
--- as it ties your application code directly to the underlying monad.
-withWeaving :: forall m a r
-            .   Member (Final m) r
-            => (forall f.
-                    Functor f
-                 => f ()
-                 -> (forall x. f (Sem r x) -> m (f x))
-                 -> (forall x. f x -> Maybe x)
-                 -> m (f a)
-               )
-            -> Sem r a
 
 -----------------------------------------------------------------------------
--- | 'withWeaving' admits an implementation of 'sendM'.
+-- | 'withWeaving' admits an implementation of 'embed'.
 --
--- Just like 'sendM', you are discouraged from using this in application code.
-embedFinal :: Functor m => Member (Final m) r => m a -> Sem r a
+-- Just like 'embed', you are discouraged from using this in application code.
+embedFinal :: (Member (Final m) r, Functor m) => m a -> Sem r a
 embedFinal m = withWeaving $ \s _ _ -> (<$ s) <$> m
-
+{-# INLINE embedFinal #-}
 
 -----------------------------------------------------------------------------
 -- | Allows for embedding higher-order actions of the final monad
--- by providing the means of explicitly threading effects through 'Sem r'
+-- by providing the means of explicitly threading effects through @'Sem' r@
 -- to the final monad. This is done through the use of the 'Strategic'
--- environment.
+-- environment, which provides 'runSemS' and 'bindSemS'.
 --
 -- You are discouraged from using 'withStrategic' in application code,
--- as it ties your application code directly to the underlying monad.
-withStrategic :: Member (Final m) r => Strategic m (Sem r) a -> Sem r a
-withStrategic strat = withWeaving $ \s wv ins -> runStrategy s wv ins strat
+-- as it ties your application code directly to the final monad.
+withStrategic :: (Member (Final m) r, Applicative m) => Strategic m (Sem r) a -> Sem r a
+withStrategic strat = withWeaving $ \s wv ins ->
+  runStrategy
+    s
+    wv
+    ins
+    strat
+{-# INLINE withStrategic #-}
 
 ------------------------------------------------------------------------------
 -- | Like 'interpretH', but may be used to
 -- interpret higher-order effects in terms of the final monad.
 --
--- /Beware/: Any interpreters built using this (or 'Final' in general)
--- will /not/ respect local/global state semantics based on the order of
--- interpreters run. You should signal interpreters that make use of
--- 'Final' by adding a "-Final" suffix to the names of these.
+-- 'interpretFinal' requires less boilerplate than using 'interpretH'
+-- together with 'withStrategic'\/'withWeaving', but is also less powerful.
+-- 'interpretFinal' does not provide any means of executing actions
+-- of @'Sem' r@ as you interpret each action, and the provided interpreter
```

Comment:
Tangentially, @ChrisPenner and I were looking for an action `Member (Final m) r => (Sem r a -> m a) -> Sem r a` the other day. Can we do that with `Final`? This comment suggests yes, but I couldn't find any actual code that did it.

Reply:
What I mean by this is that `interpretFinal` does not let you use any actions of `Sem r` in any way, shape, or form as you interpret each action. For example, you can't read an `Input` for each action, which you could do using `interpretH` + `withStrategic`:

```haskell
ex = interpretH $ \case
  Foo im -> do
    i <- input
    m <- ex (runT (im i)) 
    raise $ withStrategic $ do
      m' <- runS m
      ...
```

Reply:
```Haskell
Member (Final m) r => ((forall x. Sem r x -> m x) -> m a) -> Sem r a
```
Is not possible using `Final`; you have to use the inspector or distribution.
It's in this sense the power `Final` provides is slightly weaker than a straight-up natural transformation.

Reply:
Dang; was really hoping we'd get that by knowing the final monad.